Support atomic rename when TRUNCATE is used with INTO OUTFILE

onkar opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support atomic rename when `TRUNCATE` is used with `INTO OUTFILE`. Resolves https://github.com/ClickHouse/ClickHouse/issues/70323.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

onkar committed:
Support atomic rename when TRUNCATE is used with INTO OUTFILE

onkar committed:
Declare CANNOT_WRITE_TO_FILE inside ClientBase.cpp

onkar committed:
Make style check happy

onkar committed:
Remove duplicate/unused if block around performing atomic rename

onkar committed:
Handle special files during atomic rename

onkar committed:
Handle special files correctly

pamarcos commented on src/Client/ClientBase.cpp:
```diff
@@ -143,6 +145,78 @@ namespace ProfileEvents
+    if (!query_with_output->is_outfile_truncate)
+        return;
+
+    // Skip atomic rename for special files
```
Please use `///` for all single-line comments to follow [our coding style](https://clickhouse.com/docs/development/style#comments)

>8. Single-line comments begin with three slashes: /// and multi-line comments begin with /**. These comments are considered "documentation".

pamarcos commented on src/Client/ClientBase.cpp:
```diff
@@ -143,6 +145,78 @@ namespace ProfileEvents
+        return;
+
+    // Skip atomic rename for special files
+    if (out_file == "/dev/null" || out_file == "/dev/stdout" || out_file == "/dev/stderr")
```
Since we're checking this same thing in three different places, I think we should either move the logic to a single function, or have a common `static const std::set<String>` to check against.

pamarcos commented on src/Client/ClientBase.cpp:
```diff
@@ -143,6 +145,78 @@ namespace ProfileEvents
+                // Clean up temporary file
+                if (fs::exists(tmp_file))
+                    fs::remove(tmp_file);
+                throw DB::Exception(DB::ErrorCodes::CANNOT_WRITE_TO_FILE,
```
Please, add an empty new line before the throw to separate statements

pamarcos commented on tests/queries/0_stateless/03362_into_outfile_atomic_truncate.sh:
```diff
@@ -0,0 +1,61 @@
+    if [ "$?" -eq 0 ]; then
+        local actual
+        if [ "$is_compressed" = "1" ]; then
+            actual=$(gunzip -c "${CLICKHOUSE_TMP}/test_atomic_$test_id.out.gz")
```
Nitpick: I think it's more clear to use curly braces for variables within a string, especially if there are several variables involved:

```suggestion
            actual=$(gunzip -c "${CLICKHOUSE_TMP}/test_atomic_${test_id}.out.gz")
```

pamarcos commented on tests/queries/0_stateless/03362_into_outfile_atomic_truncate.sh:
```diff
@@ -0,0 +1,61 @@
+        if [ "$is_compressed" = "1" ]; then
+            actual=$(gunzip -c "${CLICKHOUSE_TMP}/test_atomic_$test_id.out.gz")
+        else
+            actual=$(cat "${CLICKHOUSE_TMP}/test_atomic_$test_id.out")
```
```suggestion
            actual=$(cat "${CLICKHOUSE_TMP}/test_atomic_${test_id}.out")
```

pamarcos commented on src/Client/ClientBase.cpp:
```diff
@@ -143,6 +145,78 @@ namespace ProfileEvents
+            if (tmp_file == "/dev/null" || tmp_file == "/dev/stdout" || tmp_file == "/dev/stderr")
+                return;
+
+            String out_file = tmp_file.substr(0, tmp_file.find(".tmp."));
```
We should use `rfind` instead. Otherwise, `INTO OUTFILE myfile.tmp` won't work. This edge case is a good candidate for a new test condition in the stateless test, by the way
```suggestion
            String out_file = tmp_file.substr(0, tmp_file.rfind(".tmp."));
```

pamarcos commented on tests/queries/0_stateless/03362_into_outfile_atomic_truncate.sh:
```diff
@@ -0,0 +1,61 @@
+        exit 1
+    fi
+    if [ "$is_compressed" = "1" ]; then
+        rm -f "${CLICKHOUSE_TMP}/test_atomic_$test_id.out.gz"
```
```suggestion
        rm -f "${CLICKHOUSE_TMP}/test_atomic_${test_id}.out.gz"
```

pamarcos reviewed (commented):
Great work! I think overall is en elegant solution that addresses the problem :rocket: 
I left some minor nitpick comments with what I found.

For the CHANGELOG entry I think we should write a shorter and more straightforward thing. Something more in line with the title of this PR is ok.

Also, please merge latest master against your branch, because we upgraded some submodules recently and testing on our local machines takes more time the more it deviates from master :pray:

onkar committed:
Address comments

onkar committed:
Merge remote-tracking branch 'upstream/master' into onkar/outfile-truncate-70323

onkar commented:
@pamarcos thanks for the review. I have addressed comments and merged latest `master`. Only [this test](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=77181&sha=73ef35887adb6f8dc5545c4acb975c95aeae566c&name_0=PR&name_1=Stateless+tests+%28azure%2C+asan%2C+1%2F3%29&name_2=03273_ghdata_json_to_json_alter) is failing but it is unrelated to the changes. Can you merge the PR if everything looks alright?

alexey-milovidov commented on src/Client/ClientBase.cpp:
```diff
@@ -143,6 +145,84 @@ namespace ProfileEvents
+
+bool isSpecialFile(const String & path)
+{
+    return path == "/dev/null" || path == "/dev/stdout" || path == "/dev/stderr";
```
This is wrong.

alexey-milovidov commented:
@pamarcos, this does not work at all:

```
ch "
SELECT number FROM numbers_mt(1000000) INTO OUTFILE 'numbers.txt.zst' TRUNCATE;
SELECT count() FROM file('numbers.txt.zst', 'LineAsString');
"
```

alexey-milovidov commented:
> Great work! I think overall is en elegant solution that addresses the problem

No.
